editor/editor_interface.cpp

```diff
@@ -263,6 +265,85 @@ void EditorInterface::set_current_feature_profile(const String &p_profile_name)
 	EditorFeatureProfileManager::get_singleton()->set_current_profile(p_profile_name, true);
 }
 
+// Editor dialogs.
+
+void EditorInterface::popup_dialog_scene_tree(const Callable &p_callback, const TypedArray<StringName> &p_valid_types) {
+	if (dialog_scene_tree) {
+		dialog_scene_tree->queue_free();
+	}
+	dialog_scene_tree = memnew(SceneTreeDialog);
+
+	Vector<StringName> valid_types;
+	int length = p_valid_types.size();
+	valid_types.resize(length);
+	for (int i = 0; i < length; i++) {
+		valid_types.write[i] = p_valid_types[i];
+	}
+	dialog_scene_tree->set_valid_types(valid_types);
+
+	get_base_control()->add_child(dialog_scene_tree);
+
+	dialog_scene_tree->popup_scenetree_dialog();
+
+	const Callable selected_callback = callable_mp(this, &EditorInterface::_node_selected).bind(p_callback);
+	dialog_scene_tree->connect(SNAME("selected"), selected_callback, CONNECT_DEFERRED);
+
+	const Callable canceled_callback = callable_mp(this, &EditorInterface::_node_selection_canceled).bind(p_callback);
+	dialog_scene_tree->connect(SNAME("canceled"), canceled_callback, CONNECT_DEFERRED);
+}
+
+void EditorInterface::popup_dialog_property_selector(Object *p_object, const Callable &p_callback, const Vector<int> &p_type_filter) {
+	if (dialog_property_selector) {
+		dialog_property_selector->queue_free();
```

Comment:
Any reason why you are removing the old dialog?

Reply:
I figured you'd never want multiple dialogs open at once. That said, it does seem like a fairly arbitrary limitation to impose, so I'll go ahead and remove it.

Reply:
Actually, if I don't remove the old dialog, will there be a memory leak?

I suppose the dialog will be freed when the user closes the window, but it seems dubious to me to have a dialog floating around in memory without an explicit reference to it. I could set up a container of dialogs so there is a way to keep track of all the open dialogs, but that would complicate the code for what seems to me like a pretty rare use case.

Reply:
You are adding the dialog as a child of Editor's base control. If a node is inside scene tree it won't leak.
It's fine if a hidden dialog is kept for later, our usual pattern is creating a popup on demand (e.g. context menu) and keeping it for later use to avoid subsequent initializations. It doesn't use much memory and usually if you needed it once there is high chance you will need it again.

Reply:
Ok after discussing in the chat, you could instead free the dialog in callback and here make `return` if the dialog exists (to prevent firing the dialog before the user accepts the previous one).
Or leave it as is tbh.

Reply:
It seems to add complications to the code that don't seem to provide a practical benefit. I like the idea of reusing the same popup object instead of deleting and creating a fresh one every time, but the type filter methods would have to be reworked for that approach to work. I've left a TODO in the code explaining the potential improvement.